File: src/main/java/app/application/utils/YoutubeUrlValidator.java
package app.application.utils;

import app.application.exception.InvalidPlaylistUrlException;
import app.application.exception.InvalidVideoUrlException;
import org.apache.commons.validator.routines.UrlValidator;
import org.springframework.stereotype.Service;

@Service
public class YoutubeUrlValidator {

    private final UrlValidator urlValidator = new UrlValidator();

    public void checkVideoUrl(String url) throws InvalidVideoUrlException {
        if(!urlValidator.isValid(url) && !isYoutubeUrl(url) && !hasVideoID(url)){
            throw new InvalidVideoUrlException(url);
        }
    }

    public void checkPlaylistUrl(String url) throws InvalidPlaylistUrlException {
        if(!urlValidator.isValid(url) && !hasPlaylistID(url)){
            throw new InvalidPlaylistUrlException(url);
        }
    }

    public boolean isShortUrl(String url){
        return url.contains(GlobalValues.YOUTUBE_SHORT_URL);
    }

    private boolean isYoutubeUrl(String url){
        return url.contains(GlobalValues.YOUTUBE_URL) || isShortUrl(url);
    }

    private boolean hasVideoID(String url){
        return url.contains("v=") || isShortUrl(url);
    }

    private boolean hasPlaylistID(String url){
        return url.contains("list=");
    }

}

File: src/main/kotlin/app/application/utils/YoutubeUrlValidator.kt
package app.application.utils

import kotlin.Throws
import app.application.exception.InvalidVideoUrlException
import app.application.exception.InvalidPlaylistUrlException
import app.application.utils.GlobalValues
import org.apache.commons.validator.routines.UrlValidator
import org.springframework.stereotype.Service

@Service
class YoutubeUrlValidator {

    private val urlValidator = UrlValidator()

    @Throws(InvalidVideoUrlException::class)
    fun checkVideoUrl(url: String) {
        if (!urlValidator.isValid(url) || !isYoutubeUrl(url) || !hasVideoID(url)) {
            throw InvalidVideoUrlException(url)
        }
    }

    @Throws(InvalidPlaylistUrlException::class)
    fun checkPlaylistUrl(url: String) {
        if (!urlValidator.isValid(url) || !hasPlaylistID(url)) {
            throw InvalidPlaylistUrlException(url)
        }
    }

    fun isShortUrl(url: String): Boolean {
        return url.contains(GlobalValues.YOUTUBE_SHORT_URL)
    }

    private fun isYoutubeUrl(url: String): Boolean {
        return url.contains(GlobalValues.YOUTUBE_URL) || isShortUrl(url)
    }

    private fun hasVideoID(url: String): Boolean {
        return url.contains("v=") || isShortUrl(url)
    }

    private fun hasPlaylistID(url: String): Boolean {
        return url.contains("list=")
    }
}